zson/lexer.go

```diff
@@ -343,6 +343,32 @@ func (l *Lexer) scanIdentifier() (string, error) {
 	return s, nil
 }
 
+// peekPrimitive returns a string that may be a candidate for a primitive token
+// exlusive of string literals.  This works by scanning forward until we see
+// whitespace or EOF while keeping everything buffered.  This could be made more
+// efficient by keeping track of the whitespace boundary and only looking for
+// it when we're past it (or by implementing a proper DFA for literal matching).
 func (l *Lexer) peekPrimitive() (string, error) {
+	var err error
+	var off int
+	for {
+		if off+utf8.UTFMax > len(l.cursor) {
+			err = l.check(off + utf8.UTFMax)
+			if off <= len(l.cursor) || err != nil {
+				break
+			}
+		}
```

Comment:
If the purpose of this statement is to ensure that `l.cursor[off:]`contains at least `utf8.UTFMax` bytes, then it seems like something simpler should suffice.
```suggestion
		err = l.check(off + utf8.UTFMax)
		if err != nil {
			break
		}
```